File: iris-ld/iris-ld-proofs/src/main/java/city/smartb/iris/ldproof/util/JWSUtil.java
package city.smartb.iris.ldproof.util;

import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.util.Base64URL;

import java.nio.charset.StandardCharsets;

public class JWSUtil {

	private JWSUtil() { }

	public static byte[] getJwsSigningInput(JWSHeader header, byte[] signingInput) {
		byte[] encodedHeader = encodeHeader(header);

		byte[] jwsSigningInput = new byte[encodedHeader.length + 1 + signingInput.length];
		System.arraycopy(encodedHeader, 0, jwsSigningInput, 0, encodedHeader.length);
		jwsSigningInput[encodedHeader.length] = (byte) '.';
		System.arraycopy(signingInput, 0, jwsSigningInput, encodedHeader.length+1, signingInput.length);

		return jwsSigningInput;
	}

	private static byte[] encodeHeader(JWSHeader header) {
		if (header.getParsedBase64URL() != null)
			return header.getParsedBase64URL().toString().getBytes(StandardCharsets.UTF_8);
		else
			return header.toBase64URL().toString().getBytes(StandardCharsets.UTF_8);
	}

	public static String serializeDetachedJws(JWSHeader jwsHeader, Base64URL signature) {
		return jwsHeader.toBase64URL().toString() + '.' + '.' + signature.toString();
	}
}

File: iris-ld/iris-ld-proofs/src/main/kotlin/city/smartb/iris/ld/ldproof/util/JWSUtil.kt
package city.smartb.iris.ld.ldproof.util

import com.nimbusds.jose.JWSHeader
import com.nimbusds.jose.util.Base64URL
import java.nio.charset.StandardCharsets

object JWSUtil {
    fun getJwsSigningInput(header: JWSHeader, signingInput: ByteArray): ByteArray {
        val encodedHeader = encodeHeader(header)
        val jwsSigningInput = ByteArray(encodedHeader.size + 1 + signingInput.size)
        System.arraycopy(encodedHeader, 0, jwsSigningInput, 0, encodedHeader.size)
        jwsSigningInput[encodedHeader.size] = '.'.code.toByte()
        System.arraycopy(signingInput, 0, jwsSigningInput, encodedHeader.size + 1, signingInput.size)
        return jwsSigningInput
    }

    private fun encodeHeader(header: JWSHeader): ByteArray {
        return if (header.parsedBase64URL != null) {
            header.parsedBase64URL.toString().toByteArray(StandardCharsets.UTF_8)
        } else {
            header.toBase64URL().toString().toByteArray(StandardCharsets.UTF_8)
        }
    }

    fun serializeDetachedJws(jwsHeader: JWSHeader, signature: Base64URL): String {
        return jwsHeader.toBase64URL().toString() + '.' + '.' + signature.toString()
    }
}
